src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java

```diff
@@ -856,121 +847,18 @@ private void clearActionsDone()
-     * @return true if the the item shouldn't be kept.
-     */
-    private static boolean keptEnough(@NotNull final Map<ItemStorage, Integer> alreadyKept, @NotNull final Map<ItemStorage, Integer> shouldKeep, @NotNull final ItemStack stack)
-    {
-        final ArrayList<Map.Entry<ItemStorage, Integer>> tempKeep = new ArrayList<>(shouldKeep.entrySet());
-        for (final Map.Entry<ItemStorage, Integer> tempEntry : tempKeep)
-        {
-            final ItemStorage tempStorage = tempEntry.getKey();
-            if (tempStorage != null && tempStorage.getItem() == stack.getItem() && tempStorage.getDamageValue() != stack.getItemDamage())
-            {
-                shouldKeep.put(new ItemStorage(stack, tempStorage.ignoreDamageValue()), tempEntry.getValue());
-                break;
-            }
-        }
-        final ItemStorage tempStorage = new ItemStorage(stack, false);
-
-        //Check first if the the item shouldn't be kept if it should be kept check if we already kept enough of them.
-        return shouldKeep.get(tempStorage) == null
-                || (alreadyKept.get(tempStorage) != null
-                && alreadyKept.get(tempStorage) >= shouldKeep.get(tempStorage));
-    }
-
-    /**
-     * Handle the cases when X items should be kept.
-     *
-     * @param alreadyKept already kept items.
-     * @param shouldKeep  to keep items.
-     * @param tempStorage item to analyze.
-     * @return InventoryUtils.EMPTY if should be kept entirely, else itemStack with amount which should be dumped.
-     */
-    @Nullable
-    private static ItemStack handleKeepX(
-            @NotNull final Map<ItemStorage, Integer> alreadyKept,
-            @NotNull final Map<ItemStorage, Integer> shouldKeep, @NotNull final ItemStorage tempStorage)
-    {
-        int amountKept = 0;
-        if (alreadyKept.get(tempStorage) != null)
-        {
-            amountKept = alreadyKept.remove(tempStorage);
-        }
-
-        if (shouldKeep.get(tempStorage) >= (tempStorage.getAmount() + amountKept))
-        {
-            alreadyKept.put(tempStorage, tempStorage.getAmount() + amountKept);
-            return ItemStackUtils.EMPTY;
-        }
-        alreadyKept.put(tempStorage, shouldKeep.get(tempStorage));
-        final int dump = tempStorage.getAmount() + amountKept - shouldKeep.get(tempStorage);
-
-        //Create tempStack with the amount of items that should be dumped.
-        return new ItemStack(tempStorage.getItem(), dump, tempStorage.getDamageValue());
+                || InventoryFunctions.matchFirstInHandlerWithAction(new InvWrapper(worker.getInventoryCitizen()),
+                itemStack -> !ItemStackUtils.isEmpty(itemStack) && !AbstractBuilding.buildingRequiresCertainAmountOfItem(buildingWorker, itemStack, alreadyKept),
+                (handler, slot) -> InventoryUtils.transferItemStackIntoNextFreeSlotInItemHandlers(new InvWrapper(worker.getInventoryCitizen()), slot, new InvWrapper(buildingWorker.getTileEntity()))
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Split this 197 characters long line (which is greater than 180 authorized). [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS00103)